util/genrest/resttools/error_response.go

```diff
@@ -62,18 +76,76 @@ func GRPCToHTTP(c codes.Code) int {
 	return httpStatus
 }
 
-// ErrorResponse is a helper that formats the given response information,
-// including the HTTP Status code, a message, and any error detail types, into
-// a googleAPIError and writes the response as JSON.
-func ErrorResponse(w http.ResponseWriter, status int, message string, details ...interface{}) {
-	googleAPIError := &googleAPIError{
-		Error: &googleapi.Error{
-			Code:    status,
+// HTTPToGRPC maps the given HTTP status code to the canonical gRPC
+// Code as defined by the internal source go/http-canonical-mapping.
+func HTTPToGRPC(httpStatus int) codes.Code {
+	var gRPCCode codes.Code
+	switch {
+	case httpStatus >= 200 && httpStatus < 300:
+		gRPCCode = codes.OK
+	case httpStatus >= 300 && httpStatus < 400:
+		gRPCCode = codes.Unknown
+	case httpStatus >= 400 && httpStatus < 500:
+		gRPCCode = codes.FailedPrecondition
+	case httpStatus >= 500 && httpStatus < 600:
+		gRPCCode = codes.Internal
+	default:
+		gRPCCode = codes.Unknown
+	}
+
+	if codeOverride, ok := httpToGRPC[httpStatus]; ok {
+		gRPCCode = codeOverride
+	}
+	return gRPCCode
+}
+
+// Constants to make it obvious when we're not supplying either a gRPC or HTTP status code to
+// ErrorResponse(). The code we're not supplying will be obtained from the one we do supply.
+const (
+	NoCodeGRPC codes.Code = 9999
+	NoCodeHTTP int        = -1
+)
+
+// ErrorResponse is a helper that formats the given response information, including the HTTP or gRPC
+// status code, a message, and any error detail types, into a RestError proto message and writes the
+// response as JSON.
+func ErrorResponse(w http.ResponseWriter, httpResponseCode int, grpcStatus codes.Code, message string, details ...interface{}) {
+	if httpResponseCode == NoCodeHTTP {
+		if grpcStatus == NoCodeGRPC {
+			WriteShowcaseRESTImplementationError(w, "neither HTTP code or gRPC status are provided for ErrorResponse. Exactly one must be provided.")
+			return
+		}
+		httpResponseCode = GRPCToHTTP(grpcStatus)
+	} else {
+		if grpcStatus != NoCodeGRPC {
+			WriteShowcaseRESTImplementationError(w, "both HTTP code and gRPC status are provided for ErrorResponse. Exactly one must be provided.")
+			return
+		}
+		grpcStatus = HTTPToGRPC(httpResponseCode)
+	}
+
+	jsonError := &pb.RestError{
+		Error: &pb.RestError_Status{
+			Code:    int32(httpResponseCode),
 			Message: message,
-			Details: details,
+			Status:  code.Code(grpcStatus),
 		},
 	}
-	w.WriteHeader(status)
-	data, _ := json.Marshal(googleAPIError)
+
+	for _, oneDetail := range details {
+		detailAsAny, err := anypb.New(oneDetail.(proto.Message))
+		if err != nil {
+			WriteShowcaseRESTImplementationError(w, fmt.Sprintf("could not interpret error detail as a protobuf Any: %+v", oneDetail))
+			return
+		}
+		jsonError.Error.Details = append(jsonError.Error.Details, detailAsAny)
+	}
```

Comment:
Nit: Feel free to ignore this :) I think it is just as readable with super short variable identifiers, but there is probably a middle ground. This is totally a style thing ! no right/wrong
```suggestion
	for _, d := range details {
		da, err := anypb.New(d.(proto.Message))
		if err != nil {
			WriteShowcaseRESTImplementationError(w, fmt.Sprintf("could not interpret error detail as a protobuf Any: %+v", d))
			return
		}
		jsonError.Error.Details = append(jsonError.Error.Details, da)
	}
```

Reply:
I get the argument for short identifiers, but I am generally opposed to them personally, unless maybe if they are only in really small, tight loops. And I really avoid one-character identifiers, because if there's any situation where one needs to do text-based (rather than semantic) substitution, they are pain.

Plus, I think they make the flow clearer (less cognitive load) when I'm reading the code to myself.

So I'll keep this,l if you don't mind.

Reply:
SGTM that's fine!